### PR TITLE
Dynamic Blocks and Landmines

### PR DESCRIPTION
**Pull request by MZT7**
Maps - Ability to Change Max Blocks and Landmines

### CODE REVIEW

File: convex/maps.ts [hunk 1]
```diff
@@ -193,11 +197,15 @@ export const seedMaps = internalMutation({
...
             grid: map.grid,
+            maxBlocks: 2,
+            maxLandmines: 2,
           });
         } else {
           ctx.db.insert("maps", {
             level: idx + 1,
             grid: map.grid,
+            maxBlocks: 2,
+            maxLandmines: 2,
```
Comment (delasy) [inline convex/maps.ts:208]: needs to be 0 initially

File: convex/maps.ts [hunk 2]
```diff
@@ -193,11 +197,15 @@ export const seedMaps = internalMutation({
         if (existingMap) {
           ctx.db.patch(existingMap._id, {
             grid: map.grid,
+            maxBlocks: 2,
+            maxLandmines: 2,
```
Comment (delasy) [inline convex/maps.ts:201]: needs to be 0 initially

File: app/play/[level]/page.tsx [hunk 3]
```diff
@@ -47,6 +49,8 @@ export default function PlayLevelPage({
   useEffect(() => {
     if (map) {
       setPlayerMap(map.grid.map((row) => [...row]));
+      setMaxBlocks(Number(map.maxBlocks) ?? 0);
```
Comment (delasy) [inline app/play/[level]/page.tsx:52]: are you checking anywhere that user placed more than X blocks/landmines

File: app/playground/page.tsx [hunk 4]
```diff
@@ -61,6 +63,8 @@ export default function PlaygroundPage() {
   const [userSolution, setUserSolution] = useState<string[][]>([]);
   const [visualizingUserSolution, setVisualizingUserSolution] = useState(false);
   const [openSignInModal, setOpenSignInModal] = useState(false);
+  const [blocks, setBlocks] = useState(2);
+  const [landmines, setLandmines] = useState(2);
```
Comment (delasy) [inline app/playground/page.tsx:67]: when user is playing map, are you checking anywhere that user placed more than X blocks/landmines

File: app/play/[level]/page.tsx [hunk 5]
```diff
@@ -224,14 +228,14 @@ export default function PlayLevelPage({
                   variant={placementMode === "block" ? "default" : "outline"}
                   className="h-10"
                 >
-                  Place Block ({2 - blockCount} left)
+                  Place Block ({maxBlocks - blockCount} left)
                 </Button>
                 <Button
```
Comment (delasy) [inline app/play/[level]/page.tsx:233]: if max landmines is 0 you should not display this button

File: app/play/[level]/page.tsx [hunk 6]
```diff
@@ -224,14 +228,14 @@ export default function PlayLevelPage({
                   variant={placementMode === "block" ? "default" : "outline"}
```
Comment (delasy) [inline app/play/[level]/page.tsx:228]: if zero blocks allowed then you shouldn't display this button

**Comment by delasy** [comment]
@MZT7  Thanks for your contribution ✨ This issue was one of the hardest one out there.

Though I found one thing: you are not modifying the prompt in convex/prompts. It needs to become dynamic and every time the request is made it needs to tell exact number of blocks the AI can place. NOTE: prompt does not support placing landmines so you don't need to worry about this.

I would solve it like this: have prompt as text but add a variables to it eg:
```
You can place {MAX_BLOCKS} blocks on the map.
```
And later in `models/index.ts`:
```ts
prompt.replace("{MAX_BLOCKS}", map.maxBlocks)
```

File: app/playground/page.tsx (hunk 4 above)
Comment (delasy) [reply app/playground/page.tsx:67]: you need to alter "placement mode" logic on this page

File: convex/maps.ts (hunk 2 above)
Comment (MZT7) [reply convex/maps.ts:201]: done

File: app/playground/page.tsx (hunk 4 above)
Comment (MZT7) [reply app/playground/page.tsx:67]: for the placement mode what logic do we need

**Comment by MZT7** [comment]
> @MZT7 Thanks for your contribution ✨ This issue was one of the hardest one out there.
> 
> Though I found one thing: you are not modifying the prompt in convex/prompts. It needs to become dynamic and every time the request is made it needs to tell exact number of blocks the AI can place. NOTE: prompt does not support placing landmines so you don't need to worry about this.
> 
> I would solve it like this: have prompt as text but add a variables to it eg:
> 
> ```
> You can place {MAX_BLOCKS} blocks on the map.
> ```
> 
> And later in `models/index.ts`:
> 
> ```ts
> prompt.replace("{MAX_BLOCKS}", map.maxBlocks)
> ```

For the runModel(). The "map" parameter has an argument of map.grid instead of map.
so should I change the argument to map or create another parameter for map.maxBlocks.
```ts

    const { solution, reasoning, error } = await runModel(
      args.modelId,
      map.grid,
      activePrompt.prompt,
    );
```

**Comment by delasy** [comment]
> For the runModel(). The "map" parameter has an argument of map.grid instead of map. so should I change the argument to map or create another parameter for map.maxBlocks.

would be better to just add another parameter maxBlocks to runModel function

File: app/playground/page.tsx (hunk 4 above)
Comment (delasy) [reply app/playground/page.tsx:67]: I meant this logic: https://github.com/webdevcody/survive-the-night-sim/blob/66846d409acc5c20e748cf107cf0c6483bdeafaf/app/playground/page.tsx#L330

**Comment by MZT7** [comment]
> > For the runModel(). The "map" parameter has an argument of map.grid instead of map. so should I change the argument to map or create another parameter for map.maxBlocks.
> 
> would be better to just add another parameter maxBlocks to runModel function

done this

File: models/index.ts [hunk 7]
```diff
@@ -60,7 +61,7 @@ export async function runModel(
 
   let result;
   let reasoning: string | null = null;
-
+  prompt.replace("{MAX_BLOCKS}", maxBlocks?.toString() ?? "");
```
Comment (delasy) [inline models/index.ts:64]: you need to assign this to something :)
otherwise it has no effect

Comment (MZT7) [reply models/index.ts:64]: After assigning this to a value. Is it going to be used somewhere.
```
const maxB = prompt.replace("{MAX_BLOCKS}", maxBlocks?.toString() ?? "");
```

Comment (delasy) [reply models/index.ts:64]:
```js
const modifiedPrompt = prompt.replace(...)
```
and then you pass it to every model instead of `prompt`

Comment (MZT7) [reply models/index.ts]: done

**Review by delasy — APPROVED** [review]
lgtm

**Pull request merged**